generator/src/main/java/com/graphicsfuzz/generator/mutateapi/ExpressionGenerator.java

```diff
@@ -0,0 +1,607 @@
+    );
+    final Expr condition = new BinaryExpr(new VariableIdentifierExpr("i"),
+        generateExpr(factManager, functionDefinition, stmt, conditionValue),
+        isIncrement ? BinOp.LT : BinOp.GE);
+    final Expr increment = new UnaryExpr(new VariableIdentifierExpr("i"),
+        isIncrement ? UnOp.POST_INC : UnOp.POST_DEC);
+    final Stmt body = new BlockStmt(Arrays.asList(new ExprStmt(new BinaryExpr(
+        new VariableIdentifierExpr(varName),
+        generateExpr(
+            factManager,
+            functionDefinition,
+            stmt,
+            divisorValue),
+        BinOp.ADD_ASSIGN
+    ))), false);
+    final ForStmt forStmt = new ForStmt(init, condition, increment, body);
+    functionDefinition.getBody().insertBefore(stmt, forStmt);
+
+    if (remainder > 0) {
+      final Value remainderValue = new NumericValue((BasicType) value.getType(),
+          Optional.of(remainder));
+      functionDefinition.getBody().insertBefore(stmt, new ExprStmt(new BinaryExpr(
+          new VariableIdentifierExpr(varName),
+          generateExpr(factManager, functionDefinition, stmt, remainderValue),
+          BinOp.ADD_ASSIGN
+      )));
+    }
+
+    return new VariableIdentifierExpr(varName);
+  }
+
+  private Expr generateAdditionValue(FactManager factManager,
+                                     FunctionDefinition functionDefinition,
+                                     Stmt stmt, Value value) {
+    if (!(value instanceof NumericValue)) {
+      return null;
+    }
+
+    if (value.getType() != BasicType.INT && value.getType() != BasicType.FLOAT) {
+      return null;
+    }
+
+    final Pair<Optional<Number>, Optional<Number>> pair =
+        ((NumericValue) value).getPairSum(generator);
+    return new BinaryExpr(
+        generateExpr(factManager, functionDefinition, stmt,
+            new NumericValue((BasicType) value.getType(), pair.getLeft())),
+        generateExpr(factManager, functionDefinition, stmt,
+            new NumericValue((BasicType) value.getType(), pair.getRight())),
+        BinOp.ADD
+    );
+  }
+
+  private Expr generateKnownFunctionFact(FactManager factManager,
+                                         Value value,
+                                         FunctionDefinition functionDefinition,
+                                         Stmt stmt) {
+    final Optional<List<FunctionFact>> functionFacts = globalFactManager.getFunctionFacts(value);
+    if (!functionFacts.isPresent()) {
+      return null;
+    }
+    final FunctionFact functionFact =
+        functionFacts.get().get(generator.nextInt(functionFacts.get().size()));
+    final List<Value> argValues = functionFact.getArguments();
+    final List<Expr> args = generateFunctionArgs(factManager, functionDefinition, stmt,
+        argValues);
+    return new FunctionCallExpr(functionFact.getFunctionName(), args);
+  }
+
+  private Expr generateKnownVariableFact(FactManager factManager, Value value) {
+
```

Comment:
I think you can simplify this a lot:

final List<VariableFact> availableFacts = factManager.getVariableFacts(value);
if (availableFacts.isEmpty()) {
  return null;
}
// pick a random variable fact from the list and return as you do currently.

This avoids the need for "getVariableFacts()" and the filtering out of the ones that don't match on the value.